### PR TITLE
Add Github Action workflow for CI

### LINKED ISSUE #54: Add CI Checks
This is an umbrella issue for CI tasks.
There are many checks we should make in CI

- [ ] Boilerplate header check
- [ ] Lint check
- [ ] Build
- [ ] Unit test
- [ ] Integration test

### PR DESCRIPTION
**Pull request by VanderChen**
**What type of PR is this?**
/kind enhancement

**What this PR does / why we need it**:
Add CI workflow based on GitHub Action.

**Which issue(s) this PR fixes**:
Fixes #54 including Lint check, Build, and Unit test

**Special notes for your reviewer**:
- Linter and Unit Test only work with Golang for now.
- Unit Test example needs to be added.
- Lint issues need to be fixed.

**Does this PR introduce a user-facing change?**:
NONE

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    - name: Setup Go
+      uses: actions/setup-go@v4.0.0
+      with:
+        go-version: ${{ matrix.go-version }}
+
+    - name: Install Go Dependence
+      run: |
+        go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.32
+        sudo cp $GOBIN/* /usr/bin
```
Comment (hzxuzhonghu) [inline .github/workflows/main.yml:33]: we should put it in build.sh. FYI https://github.com/kmesh-net/kmesh/pull/88/files#r1453054773

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        sudo bash ./build.sh
+    
+    # Since requiring code generated by the Proto compilation, execute go lint after building.  
+    - name: golangci-lint
+      uses: golangci/golangci-lint-action@v3.7.0
+      with:
+        args: "--out-format colored-line-number"
+        skip-pkg-cache: true
+
+    - name: Go Test
```
Comment (hzxuzhonghu) [inline .github/workflows/main.yml:47]: Separate Unit test Job from Compile

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.32
+        sudo cp $GOBIN/* /usr/bin
+    
+    - name: Build Kmesh
+      shell: bash
+      run: |
+        sudo bash ./build.sh
+    
+    # Since requiring code generated by the Proto compilation, execute go lint after building.  
+    - name: golangci-lint
```
Comment (hzxuzhonghu) [inline .github/workflows/main.yml:41]: Separate lint  Job from Compile

File: kmesh_compile_env_pre.sh [hunk 4]
```diff
@@ -14,11 +14,11 @@ function install_libboundscheck() {
 function dependency_pkg_install() {
     if command -v apt > /dev/null; then
 	    # apt install 
-	    apt-get update && apt-get install -y git make clang libbpf-dev llvm linux-tools-generic protobuf-compiler libprotobuf-dev libprotobuf-c-dev protobuf-c-compiler cmake golang
```
Comment (hzxuzhonghu) [inline kmesh_compile_env_pre.sh:17]: maybe remove it later, when we support compile before running

File: .github/workflows/main.yml (hunk 1 above)
Comment (VanderChen) [reply .github/workflows/main.yml:33]: We will remove this step, then the PR should be merged after merging PR #88.

Comment (VanderChen) [reply .github/workflows/main.yml:33]: > we should put it in build.sh. FYI https://github.com/kmesh-net/kmesh/pull/88/files#r1453054773

We will remove this step, then the PR should be merged after merging PR #88.

File: kmesh_compile_env_pre.sh (hunk 4 above)
Comment (VanderChen) [reply kmesh_compile_env_pre.sh:17]: Another installation method should be provided here. apt install could not match go 1.21 version.

File: .github/workflows/main.yml (hunk 2 above)
Comment (VanderChen) [reply .github/workflows/main.yml:47]: As mentioned in comments, UT and Lint for go require the compilation of proto.

File: .github/workflows/main.yml (hunk 3 above)
Comment (VanderChen) [reply .github/workflows/main.yml:41]: The same reason is presented above.

File: .github/workflows/main.yml (hunk 1 above)
Comment (VanderChen) [reply .github/workflows/main.yml:33]: We will remove this step, then the PR should be merged after merging PR #88.

File: .github/workflows/main.yml (hunk 2 above)
Comment (VanderChen) [reply .github/workflows/main.yml:47]: Provide some references,
- [cilium/ebpf](https://github.com/cilium/ebpf/blob/main/.github/workflows/ci.yml) gather compile and lint together
- [cilium/cilium](https://github.com/cilium/cilium/blob/main/.github/workflows/lint-go.yaml) uses a separate lint workflow.

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -0,0 +1,49 @@
+name: Kmesh CI Workflow
+
+on:
+  push:
+    branches: [ "main"]
+  pull_request:
+    branches: [ "main" ]
```
[the remark this answers is not in the document]
Comment (VanderChen) [reply .github/workflows/main.yml]: fixed it, only pull_request remains.

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+name: Kmesh CI Workflow
+
+on:
+  push:
+    branches: [ "main"]
```
[the remark this answers is not in the document]
Comment (VanderChen) [reply .github/workflows/main.yml]: fixed it, only pull_request remains.

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
+name: Kmesh CI Workflow
+
+on: push
```
Comment (hzxuzhonghu) [inline .github/workflows/main.yml:3]: pull request

File: kmesh_compile_env_pre.sh [hunk 8]
```diff
@@ -14,11 +14,11 @@ function install_libboundscheck() {
...
+	    apt-get update && apt-get install -y git make clang libbpf-dev llvm linux-tools-generic protobuf-compiler libprotobuf-dev libprotobuf-c-dev protobuf-c-compiler cmake pkg-config
```
Comment (hzxuzhonghu) [inline kmesh_compile_env_pre.sh:17]: why add `pkg-config`

File: .github/workflows/main.yml (hunk 7 above)
Comment (VanderChen) [reply .github/workflows/main.yml]: Fixed.

File: kmesh_compile_env_pre.sh (hunk 8 above)
Comment (VanderChen) [reply kmesh_compile_env_pre.sh:17]: pkg-config is unavailable on some Ubuntu distros that we have met.

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

We may separate different jobs, i am ok do that later

**Pull request merged**